Source/lighting.cpp

```diff
@@ -232,31 +235,66 @@ void DoVision(Point position, uint8_t radius, MapExplorationType doAutomap, bool
 {
 	DoVisionFlags(position, doAutomap, visible);
 
-	static const Displacement factors[] = { { 1, 1 }, { -1, 1 }, { 1, -1 }, { -1, -1 } };
-	for (auto factor : factors) {
-		for (int j = 0; j < 23; j++) {
-			int lineLen = radius - RadiusAdj[j];
-			for (int k = 0; k < lineLen; k++) {
-				Point crawl = position + VisionCrawlTable[j][k] * factor;
-				if (!InDungeonBounds(crawl))
+	// Four quadrants on a circle
+	static const Displacement quadrants[] = { { 1, 1 }, { -1, 1 }, { 1, -1 }, { -1, -1 } };
+
+	// Loop over quadrants and mirror rays for each one
+	for (auto quadrant : quadrants) {
+		// Cast a ray for a quadrant
+		for (size_t j = 0; j < std::size(VisionRays); j++) {
+			size_t rayLen = radius;
+			rayLen = std::min(rayLen, std::size(VisionRays[j]));
+			for (size_t k = 0; k < rayLen; k++) {
+				// Calculate the next point on a ray in the quadrant
+				Point rayPoint = position + VisionRays[j][k] * quadrant;
+				if (rayPoint.x == 0 && rayPoint.y == 0)
+					// No more valid points on the ray
 					break;
-				bool blockerFlag = TileHasAny(crawl, TileProperties::BlockLight);
-				bool tileOK = !blockerFlag;
-
-				if (VisionCrawlTable[j][k].deltaX > 0 && VisionCrawlTable[j][k].deltaY > 0) {
-					tileOK = tileOK || TileAllowsLight(crawl + Displacement { -factor.deltaX, 0 });
-					tileOK = tileOK || TileAllowsLight(crawl + Displacement { 0, -factor.deltaY });
-				}
-
-				if (!tileOK)
+				if (!InDungeonBounds(rayPoint))
 					break;
 
-				DoVisionFlags(crawl, doAutomap, visible);
+				bool visible = true;
+
+				//
+				// We've cast an approximated ray on an integer 2D
+				// grid, so we need to check if a ray can pass through
+				// the diagonally adjacent tiles. For example, consider
+				// this case:
+				//
+				//        #?
+				//       ↗ #
+				//     x
+				//
+				// The ray is cast from the observer 'x', and reaches
+				// the '?', but diagonally adjacent tiles '#' do not
+				// pass the light, so the '?' should not be visible
+				// for the 2D observer.
+				//
+				// The trick is to perform two additional visibility
+				// checks for the diagonally adjacent tiles, but only
+				// for the rays that are not parallel to the X or Y
+				// coordinate lines. Parallel rays, which have a 0 in
+				// one of their coordinate components, do not require
+				// any additional adjacent visibility checks, and the
+				// tile, hit by the ray, is always considered visible.
+				//
+				if (VisionRays[j][k].deltaX > 0 && VisionRays[j][k].deltaY > 0) {
+					Displacement adjacent1 = { -quadrant.deltaX, 0 };
+					Displacement adjacent2 = { 0, -quadrant.deltaY };
+
+					visible  = TileAllowsLight(rayPoint + adjacent1);

```

Comment:
Please apply project code style (run clang-format)
```suggestion
					visible = TileAllowsLight(rayPoint + adjacent1);
```